src/tools/reports.py

```diff
@@ -0,0 +1,672 @@
+
+    frameworks = obj["summarizedFrameworkInfos"]
+    for framework in frameworks:
+        name = walk_json_field_safe(framework, "frameworkName")
+        if name is None:
+            continue
+
+        finished_time = walk_json_field_safe(framework, "frameworkCompletedTimestamp") or 0
+        finished_time = int(finished_time / 1000) # yarn's time is in millisecond
+        content = json.dumps(framework)
+        result.append({"name": name, "finished_time": finished_time,
+            "content": content})
+
+    return result
+
+
+def refresh_cache(yarn_url, launcher_url):
+    db = DB()
+
+    vc_usages = get_vc_usage(yarn_url)
+    logger.info("get %d of usage from yarn", len(vc_usages))
+
+    with db.conn:
+        cursor = db.conn.cursor()
+
+        for usage in vc_usages:
+            cursor.execute("""INSERT INTO vc_usage(username,vc,cpu,mem,gpu,time)
+                            VALUES(?,?,?,?,?,?)""",
+                            (usage.user, usage.vc, usage.cpu, usage.mem,
+                                usage.gpu, usage.time))
+
+        db.conn.commit()
+
+    apps = get_yarn_apps(yarn_url)
+    logger.info("get %d of apps from yarn", len(apps))
+
+    with db.conn:
+        cursor = db.conn.cursor()
+
+        for app in apps:
+            cursor.execute("""SELECT COUNT(*) FROM apps
+                            WHERE app_id=?""",
+                            (app["app_id"],))
+            result = cursor.fetchone()
+
+            if result[0] > 0:
+                cursor.execute("""UPDATE apps SET finished_time=?, content=?
+                                WHERE app_id=?""",
+                                (app["finished_time"], app["content"], app["app_id"]))
+            else:
+                cursor.execute("""INSERT INTO apps(app_id,finished_time,content)
+                                VALUES(?,?,?)""",
+                                (app["app_id"], app["finished_time"], app["content"]))
+
+        db.conn.commit()
+
+    frameworks = get_frameworks(launcher_url)
+    logger.info("get %d of frameworks from launcher", len(frameworks))
+
+    with db.conn:
+        cursor = db.conn.cursor()
+
+        for framework in frameworks:
+            cursor.execute("""SELECT COUNT(*) FROM frameworks
+                            WHERE name=?""",
+                            (framework["name"],))
+            result = cursor.fetchone()
+
+            if result[0] > 0:
+                cursor.execute("""UPDATE frameworks SET finished_time=?, content=?
+                                WHERE name=?""",
+                                (framework["finished_time"], framework["content"], framework["name"]))
+            else:
+                cursor.execute("""INSERT INTO frameworks(name,finished_time,content)
+                                VALUES(?,?,?)""",
+                                (framework["name"], framework["finished_time"], framework["content"]))
+
+        db.conn.commit()
+
+
+def get_vc_report(days):
```

Comment:
I think passing "begin_date" and "end_date" is more flexible. Because more common request might be monthly report such as March report, not the past 30 days report.

Reply:
I'm not sure, we will delete data older than 31 days, can not generate that report. Will need to check if this is required.

Reply:
Yes, +@scarlett2018 for details.